src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -645,33 +654,54 @@ private async Task OnBatchScopeDisposedMaybeAsync(bool useAsync)
                         newSolution: solutionChanges.Solution.RemoveProjectReference(Id, projectReference));
                 }
 
-                ClearAndZeroCapacity(_projectReferencesRemovedInBatch);
-
                 // Analyzer reference adding...
                 solutionChanges.UpdateSolutionForProjectAction(
                     Id,
                     newSolution: solutionChanges.Solution.AddAnalyzerReferences(Id, _analyzersAddedInBatch.Select(a => a.GetReference())));
-                ClearAndZeroCapacity(_analyzersAddedInBatch);
 
                 // Analyzer reference removing...
                 foreach (var analyzerReference in _analyzersRemovedInBatch)
                 {
                     solutionChanges.UpdateSolutionForProjectAction(
                         Id,
                         newSolution: solutionChanges.Solution.RemoveAnalyzerReference(Id, analyzerReference.GetReference()));
-
-                    analyzerReference.Dispose();
                 }
 
-                ClearAndZeroCapacity(_analyzersRemovedInBatch);
-
                 // Other property modifications...
                 foreach (var propertyModification in _projectPropertyModificationsInBatch)
                 {
-                    propertyModification(solutionChanges);
+                    projectUpdateState = propertyModification(solutionChanges, projectUpdateState);
+                }
+
+                return projectUpdateState;
+            },
+            onAfterUpdate: (projectUpdateState) =>
+            {
+                // It is very important that these are cleared in the onAfterUpdate action passed to ApplyBatchChangeToWorkspaceMaybeAsync
```

Comment:
this is all great.  and i think i'ts much cleaner with all this contained to a single location.